src/operator/bufferCount.ts

```diff
@@ -62,19 +70,49 @@ class BufferCountOperator<T> implements Operator<T, T[]> {
  * @extends {Ignored}
  */
 class BufferCountSubscriber<T> extends Subscriber<T> {
+  private buffer: T[] = [];
+
+  constructor(destination: Subscriber<T[]>, private bufferSize: number) {
+    super(destination);
+  }
+
+  protected _next(value: T): void {
+    const buffer = this.buffer;
+
+    buffer.push(value);
+
+    if (buffer.length == this.bufferSize) {
+      this.destination.next(buffer);
+      this.buffer = [];
+    }
+  }
+
+  protected _complete(): void {
+    const buffer = this.buffer;
+    if (buffer.length > 0) {
+      this.destination.next(buffer);
+    }
+    super._complete();
+  }
+}
+
+/**
+ * We need this JSDoc comment for affecting ESDoc.
+ * @ignore
+ * @extends {Ignored}
+ */
+class BufferSkipCountSubscriber<T> extends Subscriber<T> {
   private buffers: Array<T[]> = [];
   private count: number = 0;
 
   constructor(destination: Subscriber<T[]>, private bufferSize: number, private startBufferEvery: number) {
     super(destination);
   }
 
-  protected _next(value: T) {
-    const count = this.count++;
-    const { destination, bufferSize, startBufferEvery, buffers } = this;
-    const startOn = (startBufferEvery == null) ? bufferSize : startBufferEvery;
+  protected _next(value: T): void {
+    const { bufferSize, startBufferEvery, buffers } = this;
 
-    if (count % startOn === 0) {
+    if (this.count++ % startBufferEvery === 0) {
```

Comment:
I would prefer if we don't increment inside a conditional. I understand it just fine, but I've seen it be unnecessarily confusing for many. Can we increment before we enter the conditional?